File: 230.kth-smallest-element-in-a-bst.js
// @leet start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
		var count = [0];
		var answer = [0];
		dfs(root, k, count, answer);
		return answer[0];
};
function dfs(root, k, count, answer){
		if (!root) {
			return;
		}
		dfs(root.left, k, count, answer);
		count[0]++;
		if (count[0]==k) {
			answer[0]=root.val;
				return;
		}
		dfs(root.right, k, count, answer);
}
// @leet end

File: 230.kth-smallest-element-in-a-bst.c
// @leet start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
void dfs(struct TreeNode *root, int k, int *count, int *answer) {
  if (!root)
    return;
  dfs(root->left, k, count, answer);
  (*count)++;
  if (*count == k) {
    *answer = root->val;
    return;
  }
  dfs(root->right, k, count, answer);
}
int kthSmallest(struct TreeNode *root, int k) {
  int count = 0;
  int answer = 0;
  dfs(root, k, &count, &answer);
  return answer;
}
// @leet end
